File: src/main/java/jade/GameObject.java
package jade;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class GameObject {
    private String name;
    private List<Component> components;
    public Transform transform;
    private int zIndex;

    public GameObject(String name) {
        init(name);
    }

    public void init(String name) {
        this.name = name;
        this.components = new ArrayList<>();
        this.transform = new Transform();
        this.zIndex = 0;
    }

    public void init(String name, Transform transform, int zIndex) {
        this.init(name);
        this.transform = transform;
        this.zIndex = zIndex;
    }

    public GameObject(String name, Transform transform, int zIndex) {
        init(name, transform, zIndex);
    }

    public <T extends Component> T getComponent(Class<T> componentClass) {
        for (Component c : components) {

            if (componentClass.isAssignableFrom(c.getClass())) {
                try {
                    return componentClass.cast(c);
                } catch (ClassCastException e) {
                    e.printStackTrace();
                    assert false : "Error: Casting component. ";
                }
            }
        }
        return null;
    }

    public <T extends Component> void removeComponent(Class<T> componentClass) {
        for (int i = 0; i < components.size(); i++) {
            Component c = components.get(i);
            if (componentClass.isAssignableFrom(c.getClass())) {
                components.remove(i);
                return;
            }
        }
    }

    public void addComponent(Component component) {
        components.add(component);
        component.gameObject = this;
    }

    public void update(float dt) {
//        for (int i = 0; i < components.size(); i++) {
//            components.get(i).update(dt);
//        }
        runOnAllComponents(c -> c.update(dt));
    }

    public void start() {
        runOnAllComponents(Component::start);
//    components.forEach(Component::start);
    }

    private void runOnAllComponents(Consumer<Component> action) {
        for (int i = 0; i < components.size(); i++) {
            action.accept(components.get(i));
        }
    }

    public String getName() {
        return name;
    }

    public int getzIndex() {
        return zIndex;
    }
}

File: src/main/java/jade/GameObject.kt
package jade

import components.Component
import java.util.function.Consumer

class GameObject {
    private lateinit var name: String
    private lateinit var components: MutableList<Component>
    lateinit var transform: Transform
    private var zIndex = 0
    private var uid = -1

    companion object {
        private var ID_COUNTER = 0
        fun init(maxID: Int) {
            ID_COUNTER = maxID
        }
    }


    fun init(name: String) {
        this.name = name
        components = ArrayList()
        transform = Transform()
        zIndex = 0
    }

    fun init(name: String, transform: Transform, zIndex: Int) {
        this.init(name)
        this.transform = transform
        this.zIndex = zIndex
        this.uid = ID_COUNTER++
    }

    constructor(name: String, transform: Transform, zIndex: Int) {
        init(name, transform, zIndex)
    }

    fun <T : Component?> getComponent(componentClass: Class<T>): T? {
        for (c in components) {
            if (componentClass.isAssignableFrom(c.javaClass)) {
                try {
                    return componentClass.cast(c)
                } catch (e: ClassCastException) {
                    e.printStackTrace()
                    assert(false) { "Error: Casting component. " }
                }
            }
        }
        return null
    }

    fun <T : Component?> removeComponent(componentClass: Class<T>) {
        for (i in components.indices) {
            val c = components[i]
            if (componentClass.isAssignableFrom(c.javaClass)) {
                components.removeAt(i)
                return
            }
        }
    }

    fun addComponent(component: Component) {
        component.gereateID()
        components.add(component)
        component.gameObject = this
    }

    fun update(dt: Float) {
//        for (int i = 0; i < components.size(); i++) {
//            components.get(i).update(dt);
//        }
        runOnAllComponents { c: Component -> c.update(dt) }
    }

    fun start() {
        runOnAllComponents { obj: Component -> obj.start() }
        //    components.forEach(Component::start);
    }

    fun imgui() {
        components.forEach(Component::imgui)
    }

    private fun runOnAllComponents(action: Consumer<Component>) {
        for (i in components.indices) {
            action.accept(components[i])
        }
    }

    fun getzIndex(): Int {
        return zIndex
    }

    fun getUid(): Int {
        return uid
    }

    fun getAllComponents(): MutableList<Component> {
        return components
    }
}